include/drivers/can.h

```diff
@@ -398,6 +398,207 @@ __subsystem struct can_driver_api {
+ */
+#define CAN_STATS_FORM_ERROR_INC(dev_)			\
+	STATS_INC(Z_CAN_GET_STATS(dev_), form_error)
+
+/**
+ * @brief Increment the acknowledge error counter for a CAN device
+ *
+ * The acknowledge error counter is incremented when the CAN controller does not
+ * monitor a dominant bit in the ACK slot.
+ *
+ * @param dev_ Pointer to the device structure for the driver instance.
+ */
+#define CAN_STATS_ACK_ERROR_INC(dev_)			\
+	STATS_INC(Z_CAN_GET_STATS(dev_), ack_error)
+
+/** @cond INTERNAL_HIDDEN */
+
+/**
+ * @brief Define a statically allocated and section assigned CAN device state
+ */
+#define Z_CAN_DEVICE_STATE_DEFINE(node_id, dev_name)			\
+	static struct can_device_state Z_DEVICE_STATE_NAME(dev_name)	\
+	__attribute__((__section__(".z_devstate")));
+
+/**
+ * @brief Define a CAN device init wrapper function
+ *
+ * This does device instance specific initialization of common data (such as stats)
+ * and calls the given init_fn
+ */
+#define Z_CAN_INIT_FN(dev_name, init_fn)				\
+	static inline int UTIL_CAT(dev_name, _init)(const struct device *dev) \
+	{								\
+		struct can_device_state *state =			\
+			CONTAINER_OF(dev->state, struct can_device_state, devstate); \
+		stats_init(&state->stats.s_hdr, STATS_SIZE_32, 6,	\
+			   STATS_NAME_INIT_PARMS(can));			\
+		stats_register(dev->name, &(state->stats.s_hdr));	\
+		return init_fn(dev);					\
+	}
+
+/** @endcond */
+
+/**
+ * @brief Like DEVICE_DT_DEFINE() with CAN device specifics.
+ *
+ * @details Defines a device which implements the CAN API. May generate a custom
+ * device_state container struct and init_fn wrapper when needed depending on
+ * @kconfig{CONFIG_CAN_STATS}.
+ *
+ * @param node_id   The devicetree node identifier.
+ * @param init_fn   Name of the init function of the driver.
+ * @param pm_device PM device resources reference (NULL if device does not use PM).
+ * @param data_ptr  Pointer to the device's private data.
+ * @param cfg_ptr   The address to the structure containing the configuration
+ *                  information for this instance of the driver.
+ * @param level     The initialization level. See SYS_INIT() for
+ *                  details.
+ * @param prio      Priority within the selected initialization level. See
+ *                  SYS_INIT() for details.
+ * @param api_ptr   Provides an initial pointer to the API function struct
+ *                  used by the driver. Can be NULL.
+ */
+#define CAN_DEVICE_DT_DEFINE(node_id, init_fn, pm_device,		\
+			     data_ptr, cfg_ptr, level, prio,		\
+			     api_ptr, ...)				\
+	Z_CAN_DEVICE_STATE_DEFINE(node_id, Z_DEVICE_DT_DEV_NAME(node_id)); \
+	Z_CAN_INIT_FN(Z_DEVICE_DT_DEV_NAME(node_id), init_fn)		\
+	Z_DEVICE_DEFINE(node_id, Z_DEVICE_DT_DEV_NAME(node_id),		\
+			DEVICE_DT_NAME(node_id),			\
+			&UTIL_CAT(Z_DEVICE_DT_DEV_NAME(node_id), _init), \
+			pm_device,					\
+			data_ptr, cfg_ptr, level, prio,			\
+			api_ptr,					\
+			&(Z_DEVICE_STATE_NAME(Z_DEVICE_DT_DEV_NAME(node_id)).devstate), \
+			__VA_ARGS__)
```

Comment:
I think we should probably be a bit more careful when it comes to duplicate this to all APIs that want to add stats support. It propagates device internals to other headers just for stats, IMO this doesn't scale.

Reply:
Another point is that we're introducing specialized macros (`CAN_DEVICE_DT_DEFINE`) but no specialized devices, as for example you have in Linux.

Reply:
@gmarull You approved this very same approach as the way to do this going forward in https://github.com/zephyrproject-rtos/zephyr/pull/39788#issuecomment-982848434

Reply:
> @gmarull You approved this very same approach as the way to do this going forward in [#39788 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/39788#issuecomment-982848434)

I guess it was inertia (and was only I2C at that time), would probably not give a +1 now. I don't think this is a good solution, really.

Reply:
I have added an alternative proposal here: #41616 42872

Reply:
If you consider having every driver (imagine doing this for sensors) needs to update and maintain, and be updated as more stats are added, does that scale any better? I think that's a significantly worse proposition to ask each driver to add its own unique stats. The CAN example here is a bit different than I2C in that it doesn't manipulate the stats at the API level. It could though in the future, if you add stats into the calls to send/recv can messages as mentioned in some other comments I've made here.

Reply:
That could be solved by embedding the `struct stats_can` in a `struct can_driver_data` and require that to be the first entry in the struct pointed to by `dev->data`. This is similar to how GPIO drivers handle common, required data structures (see `struct gpio_driver_data` usage).

Reply:
I'm disappointed that the previous approach was approved in both the PR and in the API meeting forum and now disliked.

I do agree now, and agreed then that copying the macro is a burden. I have some thoughts on how to rectify that and not need a custom macro wrapper here. You would instead perhaps define two things, a Z_DEVICE_STATE_TYPE and Z_DEVICE_INIT_WRAPPER prior to inclusion of device.h. Drivers would continue to use the same macros and get stats for free. So even fewer requirements on the device driver implementation side to get common data and initializer work done for stats, and perhaps other things. It could be made opt in on a per driver basis as well when needed this way.

Happy to create a PR that removes the burden of creating a DEVICE_DEFINE macro wrapper as I did initially, and showcase how this could be used to apply stats to all sensor drivers with ease.

Reply:
I did just that here, https://github.com/zephyrproject-rtos/zephyr/pull/42880 though certainly not in review ready state, removes the need for custom macros or copying macros around. Adds a couple of hooks into the device.h macros to enable it. It also renames things to effectively commit i2c to having a common context object, without needing drivers to do much besides change an include.

Reply:
I suggest we take this up in todays dev-review meeting. @teburd, @gmarull Are you able to join?

Reply:
Yes, I will join.